include/rwqueue.h

```diff
@@ -26,30 +26,44 @@
 #include <condition_variable>
 #include <deque>
 #include <mutex>
+#include <vector>
 
 template <typename T>
 class RWQueue {
 private:
 	std::deque<T> queue{}; // faster than: vector, queue, and list
-	std::mutex mutex = {};
-	std::condition_variable has_room = {};
+	std::mutex mutex                  = {};
+	std::condition_variable has_room  = {};
 	std::condition_variable has_items = {};
-	const size_t capacity = 0;
+	const size_t capacity             = 0;
+	using stride_t = typename std::vector<T>::difference_type;
 
 public:
-	RWQueue() = delete;
-	RWQueue(const RWQueue<T> &other) = delete;
-	RWQueue<T> &operator=(const RWQueue<T> &other) = delete;
+	RWQueue()                                      = delete;
+	RWQueue(const RWQueue<T>& other)               = delete;
+	RWQueue<T>& operator=(const RWQueue<T>& other) = delete;
 
 	RWQueue(size_t queue_capacity);
 
 	bool IsEmpty();
 	size_t Size();
 	size_t MaxCapacity() const;
 
-	void Enqueue(const T &item);
-	void Enqueue(T &&item); // item will be empty (moved-out) after call
+	void Enqueue(const T& item);
+
+	// items will be empty (moved-out) after call
+	void Enqueue(T&& item);
+
 	T Dequeue();
+
+	// Bulk operations can be used to move vectors of items, which also reduces
+	// the number of lock state changes.
+
+	// Enqueues or dequeues the number of requested items from the given
+	// vectors, starting with the first index. The number of requested items
+	// can exceed the capacity (the operation will be done in chunks).
+	void BulkEnqueue(std::vector<T>& from_source, const size_t num_requested);
```

Comment:
Can `from_source` be const?

Reply:
We could - but it would mean re-enabling the previous copy-by-value APIs, where each enqueue'd element is created using its copy-constructor. 

This API here is the movable version, where the queue is allowed to moved in the contents from the std::vector, without any copies being created. (in C, this would be like taking a list of pointers). 

For the MidiWork records (that we enqueue and dequeue), each of them internally has a std::vector of the message itself - so we want to avoid copying those.